### PR TITLE
Check CUDA memory pool support

### PR DESCRIPTION
**Pull request by young-developer**
Some devices dont support memory pools but still override mempool array element from nullptr to some garbage value. I added additional check from device properties.
Includes multiple GPU pool access support and memory pools access check.

### CODE REVIEW

**Review by cebtenzzre — APPROVED** [review]
This fixes the error I was seeing.

**Comment by young-developer** [comment]
@cebtenzzre Please could you check with the latest changes because device property is not in CUDA 17 and it works only for 12+.

**Comment by cebtenzzre** [comment]
That does work, but the command line I need to build unmodified llama.cpp is getting long:
```bash
cmake -B build \
  -DLLAMA_CUBLAS=ON \
  -DCMAKE_CUDA_HOST_COMPILER=gcc-12 -DCMAKE_C_COMPILER=gcc-12 -DCMAKE_CXX_COMPILER=g++-12 \
  -DLLAMA_CUDA_FORCE_MMQ=ON \
  -DCMAKE_CUDA_FLAGS='-DGGML_CUDA_FORCE_CUSTOM_MEMORY_POOL' \
&& make -C build
```
- Force gcc 12 because a recent CUDA update broke support for gcc 13
- LLAMA_CUDA_FORCE_MMQ to avoid a massive performance hit since I don't have tensor cores
- GGML_CUDA_FORCE_CUSTOM_MEMORY_POOL since one of my GPUs doesn't have memory pool support

If there is any benefit to the built-in memory pool, I'd like to have it enabled when I have my GTX 970 disabled with CUDA_VISIBLE_DEVICES. I'd also like to avoid the extra compile-time option, especially if we don't clearly document that this option is needed on older cards (MMQ isn't documented either). Could we check CUDART_VERSION and use prop.memoryPoolsSupported if available?

**Comment by young-developer** [comment]
@cebtenzzre Try without GGML_CUDA_FORCE_CUSTOM_MEMORY_POOL

**Comment by cebtenzzre** [comment]
Without that flag I get a similar error to what I was getting originally:
```
CUDA error 801 at /home/cebtenzzre/src/forks/llama.cpp/ggml-cuda.cu:6807: operation not supported
current device: 0
```

**Comment by young-developer** [comment]
So it somehow loaded the memory pool but fails on allocation to it. I can add an option based on CUDA version as you ve mentioned.

**Comment by Ph0rk0z** [comment]
I get cuda buffer pool full error when I try this out. Doesn't crash anymore but outputs one character over and over.

**Comment by young-developer** [comment]
@cebtenzzre @Ph0rk0z I added different checks(device props, alloc/dealoc test) during init phase. If one of those checks failed than it uses custom pool implementation. Please retest with new changes.

**Comment by Ph0rk0z** [comment]
Well the check works. Now both with/without GGML_CUDA_FORCE_CUSTOM_MEMORY_POOL
 scroll INCREASE_MAX_CUDA_BUFFERS at me and return one character over and over.

**Comment by young-developer** [comment]
@Ph0rk0z So it looks like it is related to another bug (not related to cuda memory pool) if it shows you the same char over and over.

**Comment by Ph0rk0z** [comment]
The PR that merged yarn caused this issue for me.

**Comment by cebtenzzre** [comment]
Commit 56e516240a923224613b132ec7062834c6485334 works fine individually on my GTX 970 and Tesla P40, but multi-GPU gives "an illegal memory access was encountered".

On 81931b2ea7729e812b3fd8cc5ff60932b0fabdb1 I get this with multi-GPU:
```
main: build = 1487 (81931b2e)
main: built with gcc-12 (GCC) 12.3.0 for x86_64-pc-linux-gnu
main: seed  = 1699115827
ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   yes
ggml_init_cublas: CUDA_USE_TENSOR_CORES: no
ggml_init_cublas: found 2 CUDA devices:
  Device 0: Tesla P40, compute capability 6.1, CUDA memory pool is supported
  Device 1: NVIDIA GeForce GTX 970, compute capability 5.2Warning: Device 1 doesnt support CUDA memory pool, skipping pool access config
Cant give access for main device memory pool to device 1

CUDA error 801 at /home/jared/src/forks/llama.cpp/ggml-cuda.cu:5929: operation not supported
current device: 0
```

File: ggml-cuda.cu [hunk 1]
```diff
@@ -5844,8 +5849,46 @@ void ggml_init_cublas() {
...
+                    err = (cudaMemPoolSetAttribute(g_cudaMemPools[id], cudaMemPoolAttrReleaseThreshold, &treshold));
+                    if (err == cudaSuccess) {
+                        fprintf(stderr, ", CUDA memory pool is supported\n");
+                    } else {
+                        g_cudaMemPools[id] = nullptr;
+                        fprintf(stderr, ", CUDA memory pool is not supported (release threshold error)\n");
+                    }
+                } else {
+                    g_cudaMemPools[id] = nullptr;
+                    fprintf(stderr, ", CUDA memory pool is not supported (cant load default pool)\n");
```
Comment (cebtenzzre) [inline ggml-cuda.cu:5872]: "can't" is misspelled in a few places:
```suggestion
                    fprintf(stderr, ", CUDA memory pool is not supported (can't load default pool)\n");
```

**Comment by cebtenzzre** [comment]
> The PR that merged yarn caused this issue for me.

If you can reproduce the problem on latest master then you should open a new issue. There were some mistakes when I merged YaRN, but CUDA and Metal seem to be working fine for most people after the fixup PRs.

File: ggml-cuda.cu (hunk 1 above)
Comment (young-developer) [reply ggml-cuda.cu]: fixed

**Comment by young-developer** [comment]
> Commit [56e5162](https://github.com/ggerganov/llama.cpp/commit/56e516240a923224613b132ec7062834c6485334) works fine individually on my GTX 970 and Tesla P40, but multi-GPU gives "an illegal memory access was encountered".
> 
> On [81931b2](https://github.com/ggerganov/llama.cpp/commit/81931b2ea7729e812b3fd8cc5ff60932b0fabdb1) I get this with multi-GPU:
> 
> ```
> main: build = 1487 (81931b2e)
> main: built with gcc-12 (GCC) 12.3.0 for x86_64-pc-linux-gnu
> main: seed  = 1699115827
> ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   yes
> ggml_init_cublas: CUDA_USE_TENSOR_CORES: no
> ggml_init_cublas: found 2 CUDA devices:
>   Device 0: Tesla P40, compute capability 6.1, CUDA memory pool is supported
>   Device 1: NVIDIA GeForce GTX 970, compute capability 5.2Warning: Device 1 doesnt support CUDA memory pool, skipping pool access config
> Cant give access for main device memory pool to device 1
> 
> CUDA error 801 at /home/jared/src/forks/llama.cpp/ggml-cuda.cu:5929: operation not supported
> current device: 0
> ```

Please retest once again.

**Comment by cebtenzzre** [comment]
> Please retest once again.

The latest commit seems to work fine with either GPU or with both. Thanks!

**Comment by young-developer** [comment]
@cebtenzzre hm, so one uses CUDA pool and another custom pool , hm interesting

**Comment by staviq** [comment]
Still broken for me

RTX 2070
plus
p106-100

@ 4ff1046d75e64f0e556d8dcd930ea25c23eb8b18
`works`
@ d6069051de7165a4e06662c89257f5d2905bb156
`CUDA error 1 at ggml-cuda.cu:7036: invalid argument`

This PR
@ 863166b4c39662b9e86406c7ee23014fb4e2d6bf
`CUDA error 217 at ggml-cuda.cu:6881: peer access is not supported between these two devices`

**Comment by cebtenzzre** [comment]
Actually, I just tested with full GPU offload instead of partial offload and got this console spam:
```
WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
...
```
So, not fixed yet.

**Comment by staviq** [comment]
@young-developer 
> I have only one GPU so it is blind fixing :)

I have my RTX2070 and p106 together in a VM, if you send me your ssh pub key and IP address I can let you in for testing, `staviq at gmail.com`

**Comment by young-developer** [comment]
I think I will change CUDA mem pool to optional because different multiple GPU can be used and will take me some time to check most cases.

**Comment by young-developer** [comment]
@cebtenzzre @staviq I added `LLAMA_CUDA_USE_CUDA_POOL`  so you can recompile if you want to test multiple GPUs using CUDA pools.

**Comment by slaren** [comment]
This is getting more complex than I would like for what is supposed to be a temporary solution, so if we have to resort to disabling this by default I would prefer to revert the original PR instead.

**Comment by cebtenzzre** [comment]
> I added `LLAMA_CUDA_USE_CUDA_POOL` so you can recompile if you want to test multiple GPUs using CUDA pools. Once it will be stable we can enable it by default.

With the latest commit I am back to this with multi-GPU:
```
CUDA error 700 at /home/jared/src/forks/llama.cpp/ggml-cuda.cu:7178: an illegal memory access was encountered
current device: 1
```

**Comment by young-developer** [comment]
> This is getting more complex than I would like for what is supposed to be a temporary solution, so if we have to resort to disabling this by default I would prefer to revert the original PR instead.

Yeap. Definitely an option.

**Comment by Ph0rk0z** [comment]
> Actually, I just tested with multi-GPU and full GPU offload instead of partial offload and got this console spam:
> 
> ```
> WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
> WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
> WARNING: cuda buffer pool full, increase MAX_CUDA_BUFFERS
> ...
> ```
> 
> So, not fixed yet.

This is the error I was talking about. Did it also output nonsense?

After yarn I had the allocation error I mentioned and the model would never load. Subsequent commits brought it to this point. I just reference it because that is the point I stopped being able to span a model across multiple GPUs.

**Comment by cebtenzzre** [comment]
> This is the error I was talking about. Did it also output nonsense?

If #3944 works for you then YaRN is no longer a problem. I can't check it right now because my GPU is busy, but I've basically reverted the same commits without issue on my local fork.

**Comment by Ph0rk0z** [comment]
It's working now finally. Also pascal prompt processing is fixed without: https://github.com/ggerganov/llama.cpp/pull/3816

Speed is mostly the same, only a few .10 difference. Not for the better but oh well.

**Pull request closed without merging**